services/can_logger/can_logger.py

```diff
@@ -154,163 +205,128 @@ def log_can(can_interface):
-        
         if(len(rx_buff) >= 1000):
-
             wr_buff.clear()
             wr_buff = rx_buff.copy()
             rx_buff.clear()
-
             if(logtodb):
-
                 p_db = mp.Process(target=write_to_db, args=(db, wr_buff,
                                                             can_bus,))
                 p_db.start()
-
             if(logtocsv):
-
                 p_csv = mp.Process(target=write_to_csv, args=(wr_buff,
                                                               can_bus,))
                 p_csv.start()
 
 # Get host info using environment variables
-
 host_ip = os.environ['socketcand_ip']
 host_port = os.environ['socketcand_port']
 host_interfaces = os.environ['can_interface']
-logging = os.environ['log']
-
+log_to = os.environ['log']
+# Check if the log rotation environment variables are present. If not, initialize
+# both period and interval with default values that will disable log rotation.
+try:
+    period = os.environ['rotation_period']
+    interval = os.environ['rotation_interval']
+except(KeyError):
+    print('WARNING: No rotation period and interval specified. Log rotation disabled')
+    period = 0
+    interval = ''
 # Split host_interfaces string into a list of strings.
-
 host_interfaces = host_interfaces.split(',')
 
 # Initialize variables
-
 can_interfaces = []
 logtodb = False
 logtocsv = False
 socket_connected = False
 db_started = False
+log_path = False
 
 # Check log selection from env. variables
-
-if (logging.find('db') != -1):
-
+if (log_to.find('db') != -1):
     logtodb = True
 
-if (logging.find('csv') != -1):
-
+if (log_to.find('csv') != -1):
     logtocsv = True
 
 # Detect available CAN interfaces
-
 avail_interfaces = detect_can_interfaces()
-
 print("Detected interfaces: " + str(avail_interfaces))
 
 # Check selected CAN interfaces in env variable are available.
-
 for i in host_interfaces:
-
     if i in avail_interfaces:
-
         can_interfaces.append(i)
-
     else:
-
         print(f'Interface {i} is not valid or is not currently available')
 
 # Initialize postgres database if database logging is enabled. The database
 # sometimes is not ready to accept connections. In that case, report the issue
 # and wait 10 seconds to try again. Keep trying until a successful connection
 # can be made.
 
-if (logtodb):
-
+if(logtodb):
     while(db_started is False):
-
         try:
-
             db = db_init()
-
         except(OperationalError):
-
-            print('Error: Database system has not been started up', 
+            print('Error: Database system has not been started up',
                   'or is starting up. Waiting...')
             time.sleep(10)
             db_started = False
-
         else:
-
             db_started = True
 
 for can_bus in can_interfaces:
     print('Creating process for', can_bus)
     mp.Process(target=log_can, args=(can_bus,)).start()
+    mp.Process(target=rotate_log, args=(can_bus, interval, period,)).start()
```

Comment:
I think this can be really simplified if you just roll the log based on the can messages timestamp -- this might even be more ideal because then CAN messages are in the correct file based on time.

For every RX'd message, compare the time to the next roll over time. If it is larger, close the current file, open a new file with the current roll over time, compute the next roll over time, and continue the processing loop. You avoid the race condition games the current design is playing with `os.rename` while still logging to the old file handle and you don't need anything `mp.Process` either.

Reply:
Absolutely. However, to implement rotation based on messages it is necessary to change the codebase of the current logger, as we are not logging by single messages, but instead, we are logging batches of thousands of messages.  This is also part of the logic that needs improvement and will be improved for a NATS-based logger.